lib/services/local/sessiontracker.go

```diff
@@ -0,0 +1,379 @@
+	sessionList                 = "list"
+	gcDelay       time.Duration = time.Minute * 5
+	retryDelay    time.Duration = time.Second
+)
+
+type sessionTracker struct {
+	bk backend.Backend
+}
+
+func NewSessionTrackerService(bk backend.Backend) (services.SessionTrackerService, error) {
+	_, err := bk.Get(context.TODO(), backend.Key(sessionPrefix, sessionList))
+	if trace.IsNotFound(err) {
+		err := createList(bk)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &sessionTracker{bk}, nil
+}
+
+func createList(bk backend.Backend) error {
+	data := []byte("[]")
+	_, err := bk.Create(context.TODO(), backend.Item{Key: backend.Key(sessionPrefix, sessionList), Value: data})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *sessionTracker) loadSession(ctx context.Context, sessionID string) (types.SessionTracker, error) {
+	sessionJSON, err := s.bk.Get(ctx, backend.Key(sessionPrefix, sessionID))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	session, err := unmarshalSession(sessionJSON.Value)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return session, nil
+}
+
+// UpdatePresence updates the presence status of a user in a session.
+func (s *sessionTracker) UpdatePresence(ctx context.Context, sessionID, user string) error {
+	sessionItem, err := s.bk.Get(ctx, backend.Key(sessionPrefix, sessionID))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	session, err := unmarshalSession(sessionItem.Value)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	session.UpdatePresence(user)
+
+	sessionJSON, err := marshalSession(session)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	item := backend.Item{Key: backend.Key(sessionPrefix, sessionID), Value: sessionJSON}
+	_, err = s.bk.CompareAndSwap(ctx, *sessionItem, item)
+	if trace.IsCompareFailed(err) {
+		log.Infof("Session resource %v presence update failed, retrying: %v", sessionID, err)
+		time.Sleep(retryDelay)
+		return s.UpdatePresence(ctx, sessionID, user)
+	}
+
+	return trace.Wrap(err)
+}
+
+// GetSessionTracker returns the current state of a session tracker for an active session.
+func (s *sessionTracker) GetSessionTracker(ctx context.Context, sessionID string) (types.SessionTracker, error) {
+	session, err := s.loadSession(ctx, sessionID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return session, nil
+}
+
+// GetActiveSessionTrackers returns a list of active session trackers.
+func (s *sessionTracker) GetActiveSessionTrackers(ctx context.Context) ([]types.SessionTracker, error) {
+	sessionList, err := s.getSessionList(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sessions := make([]types.SessionTracker, len(sessionList))
+	for i, sessionID := range sessionList {
+		session, err := s.loadSession(ctx, sessionID)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		sessions[i] = session
+	}
+
+	return sessions, nil
+}
```

Comment:
I'm not a fan of this dual system of storing a separate session list.  Separate sources of truth like this are super brittle, and loading sessions individually instead of via a range request like is done for other resources is going to result in poorer performance than letting our backends perform real range requests (potentially *hugely* increases the number of round-trip hops over the network).

Even if perf wasn't a concern, failing in the right place can cause the session list and the real collection of session tracker resources to fall out of sync.  E.g. if a tracker was deleted without being removed from the list, it would basically permanently brick this function until someone manually went in to fix the db state.